cpp/src/arrow/scalar.h

```diff
@@ -0,0 +1,143 @@
+/// an array "slot"
+struct ARROW_EXPORT Scalar {
+  /// \brief Whether the value is valid (not null) or not
+  bool is_valid;
+
+  /// \brief The type of the scalar value
+  std::shared_ptr<DataType> type;
+};
+
+/// \brief A scalar value for NullType. Never valid
+struct ARROW_EXPORT NullScalar : public Scalar {
+ public:
+  NullScalar() : Scalar{false, null()} {}
+};
+
+struct ARROW_EXPORT BooleanScalar : public Scalar {
+  bool value;
+  explicit BooleanScalar(bool value, bool is_valid = true)
+      : Scalar{is_valid, boolean()}, value(value) {}
+};
+
+template <typename Type>
+struct NumericScalar : public Scalar {
+  using T = typename Type::c_type;
+  T value;
+
+  explicit NumericScalar(T value, bool is_valid = true)
+      : Scalar{is_valid, TypeTraits<Type>::type_singleton()}, value(value) {}
+};
+
+using UInt8Scalar = NumericScalar<UInt8Type>;
+using UInt16Scalar = NumericScalar<UInt16Type>;
+using UInt32Scalar = NumericScalar<UInt32Type>;
+using UInt64Scalar = NumericScalar<UInt64Type>;
+using UInt8Scalar = NumericScalar<UInt8Type>;
+using UInt16Scalar = NumericScalar<UInt16Type>;
+using UInt32Scalar = NumericScalar<UInt32Type>;
+using UInt64Scalar = NumericScalar<UInt64Type>;
+using HalfFloatScalar = NumericScalar<HalfFloatType>;
+using FloatScalar = NumericScalar<FloatType>;
+using DoubleScalar = NumericScalar<DoubleType>;
+
+struct ARROW_EXPORT BinaryScalar : public Scalar {
+  std::shared_ptr<Buffer> value;
+  explicit BinaryScalar(const std::shared_ptr<Buffer>& value, bool is_valid = true)
+      : BinaryScalar(value, binary(), is_valid) {}
+
+ protected:
+  BinaryScalar(const std::shared_ptr<Buffer>& value,
+               const std::shared_ptr<DataType>& type, bool is_valid = true)
+      : Scalar{is_valid, type}, value(value) {}
+};
+
+struct ARROW_EXPORT FixedSizeBinaryScalar : public BinaryScalar {
+  FixedSizeBinaryScalar(const std::shared_ptr<Buffer>& value,
+                        const std::shared_ptr<DataType>& type, bool is_valid = true);
+};
+
+struct ARROW_EXPORT StringScalar : public BinaryScalar {
+  explicit StringScalar(const std::shared_ptr<Buffer>& value, bool is_valid = true)
+      : BinaryScalar(value, utf8(), is_valid) {}
+};
+
+class ARROW_EXPORT Date32Scalar : public NumericScalar<Date32Type> {
+ public:
+  using NumericScalar<Date32Type>::NumericScalar;
+};
+
+class ARROW_EXPORT Date64Scalar : public NumericScalar<Date64Type> {
+ public:
+  using NumericScalar<Date64Type>::NumericScalar;
+};
+
+class ARROW_EXPORT Time32Scalar : public NumericScalar<Time32Type> {
+ public:
+  using NumericScalar<Time32Type>::NumericScalar;
+};
+
+class ARROW_EXPORT Time64Scalar : public NumericScalar<Time64Type> {
+ public:
+  using NumericScalar<Time64Type>::NumericScalar;
+};
+
+class ARROW_EXPORT TimestampScalar : public NumericScalar<TimestampType> {
+ public:
+  using NumericScalar<TimestampType>::NumericScalar;
+};
+
+struct ARROW_EXPORT Decimal128Scalar : public Scalar {
+  std::shared_ptr<Buffer> value;
```

Comment:
Why not `char[16]`?

Reply:
If anything this should probably be `arrow::Decimal128`

Reply:
There's `uint128` in abseil.io and in the future, `int128`.

Reply:
I don't think this is worth stressing over

Reply:
:+1: